src/dbus.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright 2020, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use gio_sys;
+use glib::translate::*;
+use std::num::NonZeroU32;
+use BusNameOwnerFlags;
+use BusNameWatcherFlags;
+use BusType;
+use DBusConnection;
+
+pub struct OwnerId(NonZeroU32);
+pub struct WatcherId(NonZeroU32);
+
+fn own_closure<F>(f: F) -> glib::Closure
+where
+    F: Fn(DBusConnection, &str) + Send + Sync + 'static,
+{
+    glib::Closure::new(move |args| {
+        let mut iter = args.iter();
+        let conn = iter
+            .next()
+            .unwrap()
+            .get::<DBusConnection>()
+            .unwrap()
+            .unwrap();
+        let name = iter.next().unwrap().get::<&str>().unwrap().unwrap();
+        f(conn, name);
+        None
+    })
+}
+
+fn watch_closure<F>(f: F) -> glib::Closure
+where
+    F: Fn(DBusConnection, &str, &str) + Send + Sync + 'static,
+{
+    glib::Closure::new(move |args| {
+        let mut iter = args.iter();
+        let conn = iter
+            .next()
+            .unwrap()
+            .get::<DBusConnection>()
+            .unwrap()
+            .unwrap();
+        let name = iter.next().unwrap().get::<&str>().unwrap().unwrap();
+        let name_owner = iter.next().unwrap().get::<&str>().unwrap().unwrap();
+        f(conn, name, name_owner);
+        None
+    })
+}
+
+pub fn bus_own_name_on_connection<NameAcquired, NameLost>(
+    connection: &DBusConnection,
+    name: &str,
+    flags: BusNameOwnerFlags,
+    name_acquired_closure: NameAcquired,
+    name_lost_closure: NameLost,
```

Comment:
I would remove the `_closure` suffix here (and elsewhere), just seems a bit noisy :)